doas.c

```diff
@@ -174,11 +178,67 @@ parseconfig(const char *filename, int checkperms)
 	}
 
 	yyparse();
+	yyfn = NULL;
+
 	fclose(yyfp);
 	if (parse_errors)
 		exit(1);
 }
 
+#ifdef DOAS_CONFDIR
+static int
+isconfdir(const char *dirpath)
+{
+	struct stat sb;
+
+	if (lstat(dirpath, &sb) != 0)
+		err(1, "lstat(\"%s\")", dirpath);
+
+	if ((sb.st_mode & (S_IFMT)) == S_IFDIR)
+		return 1;
+
+	errno = ENOTDIR;
+	return 0;
+}
+
+static void
+parseconfdir(const char *dirpath, int checkperms)
+{
+	struct dirent **dirent_table;
+	size_t i, dirent_count;
+	char pathbuf[PATH_MAX];
+
+	if (!isconfdir(dirpath))
+		err(1, checkperms ? "doas is not enabled, %s" :
+		    "could not open config directory %s", dirpath);
+
+	dirent_count = scandir(dirpath, &dirent_table, NULL, alphasort);
+
+	for (i = 0; i < dirent_count; i++)
+	{
+		struct stat sb;
+		size_t pathlen;
+
+		pathlen = snprintf(pathbuf, sizeof pathbuf, "%s/%s", dirpath, dirent_table[i]->d_name);
+		free(dirent_table[i]);
+
+		/* make sure path ends in .conf */
+		if (strcmp(pathbuf + (pathlen - 5), ".conf"))
```

Comment:
I think I would check `d_name` before the snprintf and make sure to check that the file name is >5.

Reply:
nevermind about `d_name`, that would require an extra `strlen`, but I think it should be checked that pathlen - 5 >0.

Reply:
In my opinion, the current approach is more robust, as the length is *guaranteed* to always be longer than 5 (`strlen("/etc/doas.d") == 11`).

We would have to do strlen() on d_name (`d_name + (strlen(d_name) - 5)`) in order to safely check the suffix there, which just adds unnecessary complexity.  A concatenation is cheap.

Reply:
I think this is resolved by 6e76fb0.

Reply:
Yes that is good, just to make sure that a shorter part either configured or with `-C` does not do weird things.